### PR TITLE
Feat/sde

### PR DESCRIPTION
**Pull request by prakharverma**
1. Added more SDEs (Benes, Sine, Sqrt).
2. Updated SDE utils to calculate grad of KL between two SSMs and its gradient (useful for inference and learning).
3. BTDGaussian class.

### CODE REVIEW

File: markovflow/sde/sde.py [hunk 1]
```diff
@@ -136,18 +137,18 @@ class OrnsteinUhlenbeckSDE(SDE):
...
     """
 
-    def __init__(self, decay: tf.Tensor, q: tf.Tensor = tf.ones((1, 1))):
+    def __init__(self, decay: float = 1.0, q: tf.Tensor = tf.ones((1, 1)), trainable: bool = False):
         """
         Initialize the Ornstein-Uhlenbeck SDE.
 
-        :param decay: λ, a tensor with shape ``(1, 1)``.
+        :param decay: λ, a float value.
         :param q: spectral density of the Brownian motion ``(state_dim, state_dim)``.
```
Comment (vincentadam87) [inline markovflow/sde/sde.py:145]: add trainable to docstrings
maybe rename, it only applies to the decay

File: markovflow/sde/sde.py [hunk 2]
```diff
@@ -180,32 +181,35 @@ class DoubleWellSDE(SDE):
...
+    where f(x(t)) = scale * x(t) (c - x(t)^2) and the spectral density of the Brownian motion is specified by q.
     """
 
-    def __init__(self, q: tf.Tensor = tf.ones((1, 1))):
+    def __init__(self, q: tf.Tensor = tf.ones((1, 1)), scale_trainable: bool = False, c_trainable: bool = False,
+                 scale: float = 4., c: float = 1.):
         """
         Initialize the Double-Well SDE.
 
         :param q: spectral density of the Brownian motion ``(state_dim, state_dim)``.
```
Comment (vincentadam87) [inline markovflow/sde/sde.py:193]: same here with docstrings

File: markovflow/sde/sde.py [hunk 3]
```diff
@@ -217,3 +221,129 @@ def diffusion(self, x: tf.Tensor, t: tf.Tensor) -> tf.Tensor:
...
+        :param q: spectral density of the Brownian motion ``(state_dim, state_dim)``.
+        """
+        super(BenesSDE, self).__init__(state_dim=q.shape[0])
+        self.q = q
+        self.theta = tf.Variable(initial_value=theta, trainable=trainable, dtype=self.q.dtype)
+
+    def drift(self, x: tf.Tensor, t: tf.Tensor = None) -> tf.Tensor:
+        """
+        Drift of the double-well process
+        ..math:: f(x(t), t) =tanh(x(t))
```
Comment (vincentadam87) [inline markovflow/sde/sde.py]:
```suggestion
        ..math:: f(x(t), t) = tanh(x(t))
```

File: markovflow/sde/sde.py [hunk 4]
```diff
@@ -217,3 +221,129 @@ def diffusion(self, x: tf.Tensor, t: tf.Tensor) -> tf.Tensor:
...
+class SineDiffusionSDE(SDE):
+    """
+    Sine diffusion SDE represented by
+    ..math:: dx(t) = sin(x(t) - \theta) dt + dB(t), the spectral density of the Brownian motion is specified by q.
+    """
+
+    def __init__(self, theta: float = 0., q: tf.Tensor = tf.ones((1, 1)), trainable=False):
+        """
+        Initialize the SDE.
+        :param q: spectral density of the Brownian motion ``(state_dim, state_dim)``.
```
Comment (vincentadam87) [inline markovflow/sde/sde.py:287]: trainable in docstrings

File: markovflow/sde/sde.py [hunk 5]
```diff
@@ -217,3 +221,129 @@ def diffusion(self, x: tf.Tensor, t: tf.Tensor) -> tf.Tensor:
...
+class SqrtDiffusionSDE(SDE):
+    """
+    Sqrt diffusion SDE represented by
+    ..math:: dx(t) = sqrt(theta |x(t)|) dt + dB(t), the spectral density of the Brownian motion is specified by q.
+    """
+
+    def __init__(self, theta: float = 1., q: tf.Tensor = tf.ones((1, 1)), trainable=False):
+        """
+        Initialize the SDE.
+        :param q: spectral density of the Brownian motion ``(state_dim, state_dim)``.
```
Comment (vincentadam87) [inline markovflow/sde/sde.py:332]: docstrings for trainable

File: markovflow/sde/sde_utils.py [hunk 6]
```diff
@@ -55,8 +61,13 @@ def _step(current_state_time, next_state_time):
         _, t_next = next_state_time
         dt = t_next[0] - t[0]  # As time grid is homogeneous
         diffusion_term = tf.cast(l(x, t) * tf.math.sqrt(dt), dtype=DTYPE)
-        x_next = x + f(x, t) * dt + tf.squeeze(diffusion_term @ tf.random.normal(x.shape, dtype=DTYPE)[..., None],
-                                               axis=-1)
+        x_next = (
```
Comment (vincentadam87) [inline markovflow/sde/sde_utils.py:63]: yuk

File: markovflow/sde/sde_utils.py [hunk 7]
```diff
@@ -226,3 +242,219 @@ def func(x, t=None, A=A, b=b):
 
     drift_difference = 0.5 * tf.reduce_sum(drift_difference) * dt
     return drift_difference
+
+
+def gaussian_log_predictive_density(mean: tf.Tensor, chol_covariance: tf.Tensor, x: tf.Tensor) -> tf.Tensor:
+    """
+        Compute the log probability density for a Gaussian.
+    """
+    mvn = MultivariateGaussian(chol_covariance=chol_covariance)
```
Comment (vincentadam87) [inline markovflow/sde/sde_utils.py:251]: it is an odd object that first gets a cov and then a mean, but I guess there is a reason behind it

File: markovflow/sde/sde_utils.py [hunk 8]
```diff
@@ -226,3 +242,219 @@ def func(x, t=None, A=A, b=b):
 
     drift_difference = 0.5 * tf.reduce_sum(drift_difference) * dt
     return drift_difference
+
+
+def gaussian_log_predictive_density(mean: tf.Tensor, chol_covariance: tf.Tensor, x: tf.Tensor) -> tf.Tensor:
+    """
+        Compute the log probability density for a Gaussian.
```
Comment (vincentadam87) [inline markovflow/sde/sde_utils.py]:
```suggestion
        Compute the log probability density of observations x under a Gaussian distribution
        parameterized by its mean and covariance.
```

File: markovflow/sde/sde_utils.py [hunk 9]
```diff
@@ -226,3 +242,219 @@ def func(x, t=None, A=A, b=b):
...
+    """
+    The function calculates the KL divergence between two SSMs, KL[SSM-q || SSM-p].
+
+    SSM-p: x_{t+1} = f^p(x_t, t) + e^p_t ; e^p_t ~ N(0, Q_t^p)
+           f^p(x_t, t) =  A^p_t x_t + b^p_t
+
+    SSM-q: x_{t+1} = f^q(x_t, t) + e^q_t ; e^q_t ~ N(0, Q_t^q)
+           f^q(x_t, t) = A^q_t x_t + b^q_t
+
+    KL[SSM-q || SSM-p] = KL[q(x0) || p(x0)] + 0.5 * ( log(|Q_t^p|/Q_t^q) + \sum_{t=0}^{N-1} E_{q(x_{t+1}, x_t)} [||x_{t+1} - f^p(x_t, t)||^{2}_{Q_t^{p}^{-1}} - ||x_{t+1} - f^q(x_t, t)||^{2}_{Q_t^{q}^{-1}}])
```
Comment (vincentadam87) [inline markovflow/sde/sde_utils.py]:
```suggestion
    KL[SSM-q || SSM-p] = KL[q(x0) || p(x0)] + 0.5 * ( log(|Q_t^p|/Q_t^q) 
    + \sum_{t=0}^{N-1} E_{q(x_{t+1}, x_t)} [||x_{t+1} - f^p(x_t, t)||^{2}_{Q_t^{p}^{-1}} 
    - ||x_{t+1} - f^q(x_t, t)||^{2}_{Q_t^{q}^{-1}}])
```

File: markovflow/sde/sde_utils.py [hunk 10]
```diff
@@ -226,3 +242,219 @@ def func(x, t=None, A=A, b=b):
...
+
+    Rather than doing the 2D quadrature on the join distribution of q(x_{t+1}, x_t) the term is further simplified to
+    reduce to 1D quadrature.
```
Comment (vincentadam87) [inline markovflow/sde/sde_utils.py]: how?

File: markovflow/sde/sde_utils.py [hunk 11]
```diff
@@ -226,3 +242,219 @@ def func(x, t=None, A=A, b=b):
...
+
+        tf.debugging.assert_shapes(
+            [
+                (func_q_val, ("B", "N", "D")),
+                (func_p_val, ("B", "N", "D")),
+                (ssm_p_process_covar_inv[:, :, 0], ("N", "D")),
+            ]
+        )
+
+        val = (func_p_val - func_q_val) * (func_p_val - func_q_val) * ssm_p_process_covar_inv[:, :, 0]
```
Comment (vincentadam87) [inline markovflow/sde/sde_utils.py:343]: It would be good to have the equation somewhere

File: markovflow/gauss_markov.py [hunk 12]
```diff
@@ -215,3 +215,31 @@ def check_compatible(dist_1: GaussMarkovDistribution, dist_2: GaussMarkovDistrib
     tf.debugging.assert_equal(dist_1.state_dim, dist_2.state_dim)
     tf.debugging.assert_equal(dist_1.batch_shape, dist_2.batch_shape)
     tf.debugging.assert_equal(dist_1.num_transitions, dist_2.num_transitions)
+
+
+class BTDGaussian:
```
Comment (vincentadam87) [inline markovflow/gauss_markov.py]: for consistency with the rest of the toolbox,
it might be best to name this
BlockTriDiagonalGaussian

File: markovflow/sde/sde_utils.py [hunk 13]
```diff
@@ -226,3 +242,219 @@ def func(x, t=None, A=A, b=b):
...
+            (m, ("N", "D")),
+            (vals, ("N", "D")),
+        ]
+    )
+
+    vals = 0.5 * tf.reduce_sum(vals)
+    return vals
+
+
+def ssm_to_btd_nat(ssm: StateSpaceModel) -> BTDGaussian:
```
Comment (vincentadam87) [inline markovflow/sde/sde_utils.py]: How about migrating this into the SSM class as a method?
This would be pretty useful

File: markovflow/sde/sde_utils.py (hunk 9 above)
Comment (vincentadam87) [reply markovflow/sde/sde_utils.py]: missing a det on Q_q

File: markovflow/sde/sde_utils.py [hunk 14]
```diff
@@ -226,3 +242,219 @@ def func(x, t=None, A=A, b=b):
...
+
+    return nat_btd
+
+
+def SSM_KL_with_grads_wrt_exp_params(
+        ssm_q: StateSpaceModel, ssm_p: StateSpaceModel
+) -> [tf.Tensor, [tf.Tensor, tf.Tensor, tf.Tensor]]:
+    """
+    The function calculates the KL divergence between two SSMs, KL[SSM-q || SSM-p], and returns the gradients wrt the
+    expectation parameters of SSM-q.
```
Comment (vincentadam87) [inline markovflow/sde/sde_utils.py:384]: It would be good to have some of the maths here to understand the code, and where the gradient flows

File: markovflow/sde/sde_utils.py [hunk 15]
```diff
@@ -226,3 +242,219 @@ def func(x, t=None, A=A, b=b):
...
+    Note:
+        1. The function only supports SSMs with state-dim=1.
+        2. Currently, batching isn't supported.
```
Comment (vincentadam87) [inline markovflow/sde/sde_utils.py:395]: Also, here 
it seems like it is under the assumption that the diffusions terms of p and q are equal.
worth mentioning because the previous function (without gradients) was more general

File: tests/conftest.py [hunk 16]
```diff
@@ -46,3 +46,17 @@ def _batch_shape_fixture(request):
 @pytest.fixture(name="output_dim", params=[1, 2])
 def _output_dim_fixture(request):
     return request.param
+
+
+@pytest.fixture(
+    name="num_transitions", params=[5, 50, 100, 500],
```
Comment (vincentadam87) [inline tests/conftest.py]: Hum, no need to test so many cases
tests are already so slow...

File: tests/unit/test_sde.py [hunk 17]
```diff
@@ -151,7 +155,7 @@ def test_KL_sde(setup):
     Test the KL_sde function which calculates the KL divergence between two SDEs with same diffusion using
     Girsanov theorem.
```
Comment (vincentadam87) [inline tests/unit/test_sde.py:156]: The test builds two different ssms, q and p, and compares two ways to evaluate KL[q|p]

File: tests/unit/test_sde.py [hunk 18]
```diff
@@ -170,10 +174,13 @@ def test_KL_sde(setup):
...
-                                               process_covariances=tf.square(ou_sde.diffusion(m, t=None)) * dt)
+    p_ssm = state_space_model_from_covariances(
+        initial_mean=x0,
+        initial_covariance=1e-1 * tf.ones_like(x0[..., None]),
+        state_transitions=A_p,
+        state_offsets=b_p,
+        process_covariances=tf.square(ou_sde.diffusion(m, t=None)) * dt,
+    )
 
     # generate a q SSM and q SDE
```
Comment (vincentadam87) [inline tests/unit/test_sde.py:184]: so p is deterministic and q is sampled from?

File: tests/unit/test_sde.py [hunk 19]
```diff
@@ -213,14 +229,47 @@ def test_ssm_to_linear_drift(setup):
...
+        initial_mean=x0,
+        initial_covariance=1e-1 * tf.ones_like(x0[..., None]),
+        state_transitions=A_p,
+        state_offsets=b_p,
+        process_covariances=tf.square(ou_sde.diffusion(m, t=None)) * dt,
+    )
     linear_drift = LinearDrift()
     linear_drift.set_from_ssm(p_ssm, dt=dt)
 
     true_A = -1 * ou_sde.decay * tf.ones_like(m)[..., None]
```
Comment (vincentadam87) [inline tests/unit/test_sde.py:262]: this is a bit messy to read
true_A is the continuous drift while A_p is the discretized transitition of a SSM
maybe it would be good to reorganize the lines here to differentiate the two

like with docstrings
```
# continuous transition statistics
# discrete transition statistics
```

File: tests/unit/test_sde.py [hunk 20]
```diff
@@ -213,14 +229,47 @@ def test_ssm_to_linear_drift(setup):
...
     np.testing.assert_allclose(true_A, linear_drift.A)
     np.testing.assert_allclose(true_b, linear_drift.b)
+
+
+def test_SSM_KL_with_grads_wrt_exp_params(with_tf_random_seed, num_transitions):
+    """
+    Test the SSM_KL_with_grads_wrt_exp_params function which calculates the KL divergence between two SSMs and the
+    gradients of it wrt the expectation params of the first SSM.
+    """
+    dist_p, _ = StateSpaceModelBuilder(
```
Comment (vincentadam87) [inline tests/unit/test_sde.py:275]: docstrings to say this creates two different ssms
and that this construction is random

**Review by vincentadam87 — CHANGES_REQUESTED** [review]
First round of comments!
Mostly good
Some little improvements that were skipped in the rush could would help
also some clarification on what is going on at time.

File: markovflow/sde/sde_utils.py (hunk 6 above)
Comment (prakharverma) [reply markovflow/sde/sde_utils.py:63]: Ok, but what makes you go yuk?

File: markovflow/sde/sde_utils.py (hunk 7 above)
Comment (prakharverma) [reply markovflow/sde/sde_utils.py:251]: It's an object already in Markovflow, which I just wanted to use to get predictive density.

File: tests/unit/test_sde.py (hunk 18 above)
Comment (prakharverma) [reply tests/unit/test_sde.py:184]: In essence, the test function generates two SSMs from OU SDE and tests the KL value. Both the SSMs are generated by randomly initializing the decay value of the drift function.

File: markovflow/sde/sde_utils.py (hunk 13 above)
Comment (prakharverma) [reply markovflow/sde/sde_utils.py]: We would need to re-write some of the parts of the function `ssm_to_naturals` as inside the SSM class we have access to state transitions, state offsets, etc.

File: markovflow/sde/sde_utils.py (hunk 15 above)
Comment (prakharverma) [reply markovflow/sde/sde_utils.py:395]: No, both functions assume that the diffusion terms are different. Actually, the core function for KL is the function `SSM_KL_along_Gaussian_path`. The second function calculates the gradient wrt the expectation parameters, which requires writing the drift in terms of expectation parameters which is what the function does. After this, it calls the `SSM_KL_along_Gaussian_path` function.
